nautobot/core/ui/object_detail.py

```diff
@@ -378,25 +384,111 @@ def __init__(
+            add_permissions (list, optional): A list of permissions required for the "add" button to be displayed. If not provided,
+                permissions are determined by default based on the model.
+            related_field_name (str, optional): The name of the field used to filter related objects (typically for query string parameters).
         """
         self.table_key = table_key
+
         super().__init__(
             body_wrapper_template_path=body_wrapper_template_path,
             body_content_template_path=body_content_template_path,
+            header_extra_content_template_path=header_extra_content_template_path,
+            footer_content_template_path=footer_content_template_path,
             **kwargs,
         )
+        self.max_display_count = max_display_count
+        self.add_button_route = add_button_route
+        if exclude_fields and include_fields:
+            raise ValueError("You can only specify either `exclude_fields` or `include_fields`")
+        self.exclude_fields = exclude_fields
+        self.include_fields = include_fields
+        self.related_field_name = related_field_name
+        self.add_permissions = add_permissions
+
+    def _get_table_add_url(self, context):
+        obj = context["obj"]
+        body_content_table_add_url = None
+        request = context["request"]
+        related_field_name = self.related_field_name or obj._meta.model_name
+        return_url = context["return_url"]
+
+        if self.add_button_route == "default":
+            body_content_table = context[self.table_key]
+
+            body_content_table_model = body_content_table.Meta.model
+            permission_name = get_permission_for_model(body_content_table_model, "add")
+            if request.user.has_perms([permission_name]):
+                with contextlib.suppress(NoReverseMatch):
+                    add_route = reverse(get_route_for_model(body_content_table_model, "add"))
+                    body_content_table_add_url = f"{add_route}?{related_field_name}={obj.pk}&return_url={return_url}"
+
+        elif self.add_button_route not in [None, "default"]:
+            if request.user.has_perms(self.add_permissions or []):
+                add_route = reverse(self.add_button_route)
+                body_content_table_add_url = f"{add_route}?{related_field_name}={obj.pk}&return_url={return_url}"
+
+        return body_content_table_add_url
 
     def get_extra_context(self, context):
         """Take the context value under `self.table_key` and add it to the context under key `"body_content_table"`."""
-        return {"body_content_table": context[self.table_key]}
+        body_content_table = context[self.table_key]
+        request = context["request"]
+
+        if self.exclude_fields is not None:
+            body_content_table.exclude = self.exclude_fields
+        elif self.include_fields is not None:
+            body_content_table.Meta.default_columns = self.include_fields
```

Comment:
Is this the best set of APIs to use, or should it instead be something like:

```py
if self.exclude_fields or self.include_fields:
    for column in body_content_table.columns:
        if (exclude_fields and column.name in exclude_fields) or (include_fields and column.name not in include_fields):
            body_content_table.columns.hide(column.name)
        else:
            body_content_table.columns.show(column.name)
```

Alternately, maybe we should change up how things work in the **view** `get_extra_context` and have it provide just the table class and the queryset in its context, and let this actually initialize the table instance?

Reply:
> Alternately, maybe we should change up how things work in the view get_extra_context and have it provide just the table class and the queryset in its context, and let this actually initialize the table instance?

Thoughts on this idea? The current pattern feels like duplicate code to me since we have to create and configure the table in the view code, then this code here overrides that with its own configuration.

Reply:
This would require me to also to modify `associated_contacts_table` in `nautobot.core.views.utils.common_detail_view_context` to supply both the `AssociatedContactsTable ` and its `queryset`. This would likely involve updates in other areas, such as the `NautobotHTMLRenderer`. 
Additionally, we can’t specify fields to include at table initialization—only fields to exclude—so we might end up using a similar approach to the current solution i.e `body_content_table.columns.show(column.name)`.

Given this, I believe the current implementation is simple and sufficient for now.

Reply:
Fair. I still feel the current pattern involves more redundancy than is desirable, but I'll open a follow up issue rather than holding this one up any further.